File: app_project/backend/app/services/news/document_service.py
from sqlalchemy.orm import Session
from fastapi import HTTPException
from app.models.news.document import Document, DocumentCategory
from app.schemas.news.document import DocumentCreate, DocumentUpdate, DocumentCategoryCreate, DocumentCategoryUpdate

# DocumentCategory
def create_category(db: Session, data: DocumentCategoryCreate) -> DocumentCategory:
    category = DocumentCategory(**data.model_dump())
    db.add(category)
    db.commit()
    db.refresh(category)
    return category

def update_category(db: Session, category_id: int, data: DocumentCategoryUpdate) -> DocumentCategory:
    category = db.query(DocumentCategory).filter(DocumentCategory.id == category_id).first()
    if not category:
        raise HTTPException(404, f"Category {category_id} not found")
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(category, field, value)
    db.commit()
    db.refresh(category)
    return category

# Document
def create_document(db: Session, data: DocumentCreate) -> Document:
    document = Document(**data.model_dump())
    db.add(document)
    db.commit()
    db.refresh(document)
    return document

def update_document(db: Session, document_id: int, data: DocumentUpdate) -> Document:
    document = db.query(Document).filter(Document.id == document_id).first()
    if not document:
        raise HTTPException(404, f"Document {document_id} not found")
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(document, field, value)
    db.commit()
    db.refresh(document)
    return document

def delete_document(db: Session, document_id: int):
    document = db.query(Document).filter(Document.id == document_id).first()
    if not document:
        raise HTTPException(404, f"Document {document_id} not found")
    db.delete(document)
    db.commit()
    return True

File: app_project/frontend/src/services/news/documentService.ts
import apiClient from "@/utils/apiClient";
import { getMultipartHeaders } from "@/utils/getMultipartHeaders";
import type {
  DocumentCreate,
  DocumentUpdate,
  DocumentResponse,
  DocumentCategoryCreate,
  DocumentCategoryUpdate,
  DocumentCategoryResponse,
} from "@/models/news/document";

const buildFormData = (data: Record<string, any>) => {
  const formData = new FormData();
  for (const key in data) {
    if (data[key] !== undefined && data[key] !== null) {
      formData.append(key, data[key]);
    }
  }
  return formData;
};

// ========== DOCUMENT ==========

export const fetchDocuments = async (): Promise<{
  success: boolean;
  data?: DocumentResponse[];
  message: string;
}> => {
  try {
    const response = await apiClient.get("/documents", getMultipartHeaders());
    return {
      success: true,
      data: response.data,
      message: "Fetched documents successfully",
    };
  } catch (error: any) {
    return {
      success: false,
      message: error.response?.data?.detail || "Failed to fetch documents",
    };
  }
};

export const getDocumentDetail = async (id: number) => {
  try {
    const response = await apiClient.get(`/documents/${id}`, getMultipartHeaders());
    return { success: true, data: response.data };
  } catch (error: any) {
    return {
      success: false,
      message: error.response?.data?.detail || "Failed to get document",
    };
  }
};

export const createDocument = async (data: DocumentCreate) => {
  try {
    const formData = buildFormData(data);
    const response = await apiClient.post("/documents", formData, getMultipartHeaders());
    return { success: true, data: response.data, message: "Document created successfully" };
  } catch (error: any) {
    return {
      success: false,
      message: error.response?.data?.detail || "Failed to create document",
    };
  }
};

export const updateDocument = async (id: number, data: DocumentUpdate) => {
  try {
    const formData = buildFormData(data);
    const response = await apiClient.put(`/documents/${id}`, formData, getMultipartHeaders());
    return { success: true, data: response.data, message: "Document updated successfully" };
  } catch (error: any) {
    return {
      success: false,
      message: error.response?.data?.detail || "Failed to update document",
    };
  }
};

export const deleteDocument = async (id: number) => {
  try {
    await apiClient.delete(`/documents/${id}`, getMultipartHeaders());
    return { success: true, message: "Document deleted successfully" };
  } catch (error: any) {
    return {
      success: false,
      message: error.response?.data?.detail || "Failed to delete document",
    };
  }
};

// ========== CATEGORY ==========

export const fetchDocumentCategories = async (): Promise<{
  success: boolean;
  data?: DocumentCategoryResponse[];
  message: string;
}> => {
  try {
    const response = await apiClient.get("/documents/categories", getMultipartHeaders());
    return {
      success: true,
      data: response.data,
      message: "Fetched categories successfully",
    };
  } catch (error: any) {
    return {
      success: false,
      message: error.response?.data?.detail || "Failed to fetch categories",
    };
  }
};

export const getCategoryDetail = async (id: number) => {
  try {
    const response = await apiClient.get(`/documents/categories/${id}`, getMultipartHeaders());
    return { success: true, data: response.data };
  } catch (error: any) {
    return {
      success: false,
      message: error.response?.data?.detail || "Failed to get category",
    };
  }
};

export const createCategory = async (data: DocumentCategoryCreate) => {
  try {
    const formData = buildFormData(data);
    const response = await apiClient.post("/documents/categories", formData, getMultipartHeaders());
    return {
      success: true,
      data: response.data,
      message: "Category created successfully",
    };
  } catch (error: any) {
    return {
      success: false,
      message: error.response?.data?.detail || "Failed to create category",
    };
  }
};

export const updateCategory = async (id: number, data: DocumentCategoryUpdate) => {
  try {
    const formData = buildFormData(data);
    const response = await apiClient.put(`/documents/categories/${id}`, formData, getMultipartHeaders());
    return {
      success: true,
      data: response.data,
      message: "Category updated successfully",
    };
  } catch (error: any) {
    return {
      success: false,
      message: error.response?.data?.detail || "Failed to update category",
    };
  }
};
